[serve] Constructor failure logic doesn't handle user_config update exceptions

```python
import ray
from ray import serve

serve.start()

@serve.deployment
class A:
    def reconfigure(self, config):
        if config == "hi":
            raise Exception("oops")

A.options(user_config="not_hi").deploy()
A.options(user_config="hi").deploy()
```

Gets into a bad infinite loop:
```
(pid=99788) 2021-08-31 15:45:57,213     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
(pid=99788) 2021-08-31 15:45:57,318     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
(pid=99788) 2021-08-31 15:45:57,423     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
(pid=99788) 2021-08-31 15:45:57,530     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
(pid=99788) 2021-08-31 15:45:57,638     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:57,745   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:57,853   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:57,957   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
(pid=99788) 2021-08-31 15:45:58,063     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:58,174   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:58,300   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:58,417   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:58,524   INFO backend_state.py:939 -- Adding 1 replicas to deployment 'A'. component=serve deployment=A
(pid=99788) 2021-08-31 15:45:58,537     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
(pid=99788) 2021-08-31 15:45:58,641     ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:58,746   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
^C(pid=99788) 2021-08-31 15:45:58,853   ERROR controller.py:124 -- Exception updating backend state: 'BackendState' object has no attribute '_backend_matadata'
```

@edoakes I just checked on master and it's failed as expected on my end: 

Also i didn't find any "_backend_matadata" usage on ray master branch anymore, is it a transient issue during refactoring ?

```
❯ pytest test_deploy.py::test_redeploy_user_config -s
Test session starts (platform: darwin, Python 3.8.10, pytest 5.4.3, pytest-sugar 0.9.4)
rootdir: /Users/jiaodong/Workspace/ray/python
plugins: sugar-0.9.4, asyncio-0.15.1, timeout-1.4.2, lazy-fixture-0.6.3, rerunfailures-10.0
collecting ... 2021-09-12 21:44:34,797	INFO services.py:1250 -- View the Ray dashboard at http://127.0.0.1:8265
(pid=4501) 2021-09-12 21:44:37,051	INFO http_state.py:75 -- Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-node:192.168.86.154-0' on node 'node:192.168.86.154-0' listening on '127.0.0.1:8000'
2021-09-12 21:44:37,288	INFO api.py:451 -- Started detached Serve instance in namespace 'default_test_namespace'.
2021-09-12 21:44:37,294	INFO api.py:241 -- Updating deployment 'A'. component=serve deployment=A
(pid=4504) INFO:     Started server process [4504]
(pid=4501) 2021-09-12 21:44:37,372	INFO backend_state.py:865 -- Adding 1 replicas to deployment 'A'. component=serve deployment=A
2021-09-12 21:44:37,701	INFO api.py:248 -- Deployment 'A' is ready at `http://127.0.0.1:8000/A`. component=serve deployment=A
2021-09-12 21:44:37,710	INFO api.py:241 -- Updating deployment 'A'. component=serve deployment=A
(pid=4501) 2021-09-12 21:44:37,803	INFO backend_state.py:817 -- Stopping 1 replicas of deployment 'A' with outdated versions. component=serve deployment=A
(pid=4501) 2021-09-12 21:44:39,940	INFO backend_state.py:865 -- Adding 1 replicas to deployment 'A'. component=serve deployment=A
(pid=4501) 2021-09-12 21:44:40,602	INFO backend_state.py:865 -- Adding 1 replicas to deployment 'A'. component=serve deployment=A
(pid=4501) 2021-09-12 21:44:41,251	INFO backend_state.py:865 -- Adding 1 replicas to deployment 'A'. component=serve deployment=A


――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_redeploy_user_config ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

serve_instance = <ray.serve.api.Client object at 0x7fb3902f04f0>

    def test_redeploy_user_config(serve_instance):
        @serve.deployment
        class A:
            def reconfigure(self, config):
                if config == "hi":
                    raise Exception("oops")

        A.options(user_config="not_hi").deploy()
>       A.options(user_config="hi").deploy()

test_deploy.py:1028:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../api.py:747: in deploy
    return _get_global_client().deploy(
../api.py:91: in check
    return f(self, *args, **kwargs)
../api.py:247: in deploy
    self._wait_for_goal(goal_id)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ray.serve.api.Client object at 0x7fb3902f04f0>, goal_id = UUID('49414d91-e739-4ee8-8587-597572a399a9'), timeout = None

    def _wait_for_goal(self,
                       goal_id: Optional[GoalId],
                       timeout: Optional[float] = None) -> bool:
        if goal_id is None:
            return True

        ready, _ = ray.wait(
            [self._controller.wait_for_goal.remote(goal_id)], timeout=timeout)
        # AsyncGoal could return exception if set, ray.get()
        # retrieves and throws it to user code explicitly.
        if len(ready) == 1:
            async_goal_exception = ray.get(ready)[0]
            if async_goal_exception is not None:
>               raise async_goal_exception
E               RuntimeError: Deployment 'A' failed, rolling back to version {rollback_info.version} asynchronously.

../api.py:181: RuntimeError

 ray/serve/tests/test_deploy.py ⨯                                                                                                                                                                             100% ██████████
================================================================================================== short test summary info ==================================================================================================
FAILED test_deploy.py::test_redeploy_user_config - RuntimeError: Deployment 'A' failed, rolling back to version {rollback_info.version} asynchronously.

Results (9.18s):
       1 failed
         - ray/serve/tests/test_deploy.py:1020 test_redeploy_user_config
```

--- a/python/ray/serve/tests/test_deploy.py
+++ b/python/ray/serve/tests/test_deploy.py
@@ -251,6 +251,34 @@ def test_config_change(serve_instance, use_handle):
 
 
 @pytest.mark.skipif(sys.platform == "win32", reason="Failing on Windows.")
+def test_reconfigure_with_exception(serve_instance):
+    @serve.deployment
+    class A:
+        def __init__(self):
+            self.config = "yoo"
+
+        def reconfigure(self, config):
+            if config == "hi":
+                raise Exception("oops")
+
+            self.config = config
+
+        def __call__(self, *args):
+            return self.config
+
+    A.options(user_config="not_hi").deploy()
+    config = ray.get(A.get_handle().remote())
+    assert config == "not_hi"
+
+    with pytest.raises(RuntimeError):
+        A.options(user_config="hi").deploy()
+
+    # Ensure we should be able to rollback to "hi" config
+    config = ray.get(A.get_handle().remote())
+    assert config == "not_hi"
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="Failing on Windows.")
 @pytest.mark.parametrize("use_handle", [True, False])
 def test_redeploy_single_replica(serve_instance, use_handle):
     # Tests that redeploying a deployment with a single replica waits for the

